In [1]:
# Analiz və vizualizasiya üçün kitabxanaların yüklənməsi
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [2]:
from google.colab import files
uploaded = files.upload()

# Datasetin yüklənməsi
df=pd.read_csv('house_sale.csv',sep=',')

Saving house_sale.csv to house_sale (7).csv


# Datasetə ümumi baxış

In [3]:
df.head()
#ilk 5 sətir

,id_x,rel_url,estate_rel_url_x,datetime_scrape_x,price,currency_x,location,attributes,city_when,city,...,Mərtəbə,Otaq sayı,Sahə,Torpaq sahəsi,Təmir,Çıxarış,İpoteka,estate_details_id_y,estate_rel_url,extra_info
0,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,/alqi-satqi?page=174,/items/4521724,2024-10-05 22:07:37.60613+00,499999.0,AZN,Səbail r.,"4 otaqlı, 145 m², 7/9 mərtəbə","Bakı, dünən 23:52",bakı,...,7 / 9,4.0,145 m²,NaN,var,var,NaN,92e82ea2-e1f3-4c2d-a284-151efd99281e,/items/4521724,Şəhidlər xiyabanı * Dağüstü parkı * Səbail r.
1,883e20f0-8872-49a5-8b4b-63b8301b5f8f,/alqi-satqi?page=250,/items/4669294,2024-10-05 22:07:37.60613+00,77000.0,AZN,Biləcəri q.,"4 otaqlı, 90 m²","Bakı, dünən 23:56",bakı,...,NaN,4.0,90 m²,1.3 sot,var,yoxdur,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,/items/4669294,Binəqədi r.* Biləcəri q.
2,55c36fb1-a3af-476e-ba17-a81f6795be8d,/alqi-satqi?page=250,/items/4669293,2024-10-05 22:07:37.60613+00,92000.0,AZN,İnşaatçılar m.,"3 otaqlı, 60 m²","Bakı, dünən 23:55",bakı,...,NaN,3.0,60 m²,0.1 sot,var,var,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,/items/4669293,İnşaatçılar m.* Yasamal r.
3,acf1aa8d-a46a-40f5-b6f2-f7a449569337,/alqi-satqi?page=250,/items/4647811,2024-10-05 22:07:37.60613+00,95000.0,AZN,Qaraçuxur q.,130 m²,"Bakı, dünən 23:55",bakı,...,NaN,NaN,130 m²,NaN,var,var,var,5db56980-05cc-4925-b55b-f58fb3f4d2b6,/items/4647811,Suraxanı r.* Qaraçuxur q.
4,22d840df-9283-4112-bc71-7432511fc776,/alqi-satqi?page=250,/items/4638863,2024-10-05 22:07:37.60613+00,220000.0,AZN,Əhmədli m.,"3 otaqlı, 100 m², 15/16 mərtəbə","Bakı, dünən 23:52",bakı,...,15 / 16,3.0,100 m²,NaN,var,var,NaN,d71cf9a9-86dd-4162-b540-6252a8659a09,/items/4638863,Əhmədli m.* Xalqlar Dostluğu m.* Xətai r.* Əhm...


In [4]:
df.shape
#datasetin ölçüsü (sətir,sütun)

(100775, 51)

Dataset 51 sütundan və 100775 sətirdən ibarətdir.

# Sütun başlıqlarına və dəyərlərinə baxış

In [5]:
df.columns
#sütun başlıqlarına ümumi baxış

Index(['id_x', 'rel_url', 'estate_rel_url_x', 'datetime_scrape_x', 'price',
       'currency_x', 'location', 'attributes', 'city_when', 'city', 'day_x',
       'hour_x', 'repair', 'vip', 'featured', 'products_label', 'bill_of_sale',
       'mortgage', 'img_url', 'id_y', 'estate_id', 'estate_rel_url_y',
       'datetime_scrape_y', 'description', 'unit_price', 'total_price',
       'currency_y', 'owner_name', 'owner_title', 'shop_name', 'shop_title',
       'address', 'lat', 'lng', 'updated', 'views', 'day_y', 'hour_y',
       'estate_details_id_x', 'Binanın növü', 'Kateqoriya', 'Mərtəbə',
       'Otaq sayı', 'Sahə', 'Torpaq sahəsi', 'Təmir', 'Çıxarış', 'İpoteka',
       'estate_details_id_y', 'estate_rel_url', 'extra_info'],
      dtype='object')

Datasetdə ümumi olaraq datalar azərbaycanca yazılıb, lakin bəzi sütun başlıqları ingiliscə adlandırılıb. Həmin sütun adlarını azərbaycanca yazıb datanı daha təmiz hala gətirməliyik. Amma əvvəlcə prosesi daha da aydınlaşdırmaq üçün sütunda "x" və "y" ilə fərqləndirilən sütunların eyni olub-olmamasını yoxlayaq.

In [6]:
id=(df['id_x'] == df['id_y']).all()
estate_rel_url=(df['estate_rel_url_x'] == df['estate_rel_url_y']).all()
datetime_scrape=(df['datetime_scrape_x'] == df['datetime_scrape_y']).all()
currency=(df['currency_x'] == df['currency_y']).all()
day=(df['day_x'] == df['day_y']).all()
hour=(df['hour_x'] == df['hour_y']).all()
details=(df['estate_details_id_x'] == df['estate_details_id_y']).all()

print(f"id: {id}")
print(f"estate_rel_url: {estate_rel_url}")
print(f"datetime_scrape: {datetime_scrape}")
print(f"currency: {currency}")
print(f"day: {day}")
print(f"hour: {hour}")
print(f"details: {details}")

id: False
estate_rel_url: True
datetime_scrape: False
currency: True
day: False
hour: False
details: True


Biraz daha araşdıraq:

In [7]:
df[['id_x','id_y','estate_rel_url_x','estate_rel_url_y','datetime_scrape_x','datetime_scrape_y','currency_x','currency_y','day_x','day_y','hour_x','hour_y','estate_details_id_x','estate_details_id_y']].head()

,id_x,id_y,estate_rel_url_x,estate_rel_url_y,datetime_scrape_x,datetime_scrape_y,currency_x,currency_y,day_x,day_y,hour_x,hour_y,estate_details_id_x,estate_details_id_y
0,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,92e82ea2-e1f3-4c2d-a284-151efd99281e,/items/4521724,/items/4521724,2024-10-05 22:07:37.60613+00,2024-10-05 22:14:10.089116+00,AZN,AZN,05.10.2024,05.10.2024,23:52,23:52,92e82ea2-e1f3-4c2d-a284-151efd99281e,92e82ea2-e1f3-4c2d-a284-151efd99281e
1,883e20f0-8872-49a5-8b4b-63b8301b5f8f,505eaf81-6bc8-4094-9b00-aa82066548ee,/items/4669294,/items/4669294,2024-10-05 22:07:37.60613+00,2024-10-05 22:14:10.089116+00,AZN,AZN,05.10.2024,04.10.2024,23:56,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,505eaf81-6bc8-4094-9b00-aa82066548ee
2,55c36fb1-a3af-476e-ba17-a81f6795be8d,fa63b201-999d-43b5-a61a-778d9d79a6c6,/items/4669293,/items/4669293,2024-10-05 22:07:37.60613+00,2024-10-05 22:14:10.089116+00,AZN,AZN,05.10.2024,04.10.2024,23:55,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,fa63b201-999d-43b5-a61a-778d9d79a6c6
3,acf1aa8d-a46a-40f5-b6f2-f7a449569337,5db56980-05cc-4925-b55b-f58fb3f4d2b6,/items/4647811,/items/4647811,2024-10-05 22:07:37.60613+00,2024-10-05 22:14:10.089116+00,AZN,AZN,05.10.2024,04.10.2024,23:55,NaN,5db56980-05cc-4925-b55b-f58fb3f4d2b6,5db56980-05cc-4925-b55b-f58fb3f4d2b6
4,22d840df-9283-4112-bc71-7432511fc776,d71cf9a9-86dd-4162-b540-6252a8659a09,/items/4638863,/items/4638863,2024-10-05 22:07:37.60613+00,2024-10-05 22:14:10.089116+00,AZN,AZN,05.10.2024,04.10.2024,23:52,NaN,d71cf9a9-86dd-4162-b540-6252a8659a09,d71cf9a9-86dd-4162-b540-6252a8659a09


ID'lərin və tarixlərin fərqli olmasına əsasən deyə bilərik ki, əmlakın dataları fərqli günlərdə yenidən toplanaraq update olunub, ona görə də estate_rel_url və estate_details_id eyni qalır, currency isə sabit bir dəyərdir və True verir.

In [8]:
x=(df['estate_rel_url'] == df['estate_rel_url_x']).all()
y=(df['estate_rel_url'] == df['estate_rel_url_y']).all()

print(f"Əmlak_URL_1: {x}")
print(f"Əmlak_URL_2: {y}")

Əmlak_URL_1: True
Əmlak_URL_2: True


Son olaraq estate_rel_url sütununa baxırıq, bu 3 sütun ayrı-ayrılıqda eyni məlumatları saxlayır. Eyni sətirdə bir məlumat 3 yerdə təkrarlanır ona görə bu sütunlardan birini saxlayıb digərlərini silirik. Estate_details_is sütunu da duplicate sütun olduğuna görə birini silirik. Currency sütununa gəldikdə isə data azərbaycanca yazıldığına görə valyuta avtomatik olaraq AZN olur və sabit qalır, ona görə də duplicate sütun hesab olunmur.

In [9]:
df.drop(columns=['estate_rel_url_y', 'estate_rel_url','estate_details_id_y'], inplace=True)

Currency sütununa daha dərindən baxsaq:

In [10]:
(df['price'] == df['total_price']).all()

np.False_

Currency (valyuta) sütununun 2 yerdə fərqli adla aid olduğu price və total price bir-birindən fərqləndiyi üçün update olunmuş datada əmlakın qiymətinin dəyişdiyi müşahidə olunur, ona görə də currency sütunu olduğu kimi qalır.

Digər diqqət çəkən sütun isə city_when sütunudur. Çünki bu sütun city, day və hour olaraq ayrıca sütunlarda bölünüb və daha aydın şəkildə əks olunub. City hissəsi eyni qalsa da, when hissəsi ilkin tarixi göstərir (avtomatik olaraq bütün tarixlər "dünən" olaraq qeyd edilib) və update tarixini əks etdirməyərək conflict yaradır. Ona görə də city_when sütunu duplicate sütun hesab olunur və silinir.

In [11]:
df.drop(columns=['city_when'], inplace=True)

"bill_of_sale" sütunu "çıxarış" olaraq adlanan sütunun ingiliscə adlandırılan versiyasıdır, hər iki sütun əmlakın çıxarışının olub-olmamasını əks etdirir. Sütunlara daha dəqiq baxsaq:

In [12]:
df['bill_of_sale'].value_counts(dropna=False)

,count
bill_of_sale,
Çıxarış var,79565
NaN,21210


In [13]:
df['Çıxarış'].value_counts(dropna=False)

,count
Çıxarış,
var,79565
yoxdur,21210


Dəyərlərə baxdıqda bill_of_sale sütununda "NaN" olaraq qeyd olunanların, Çıxarış sütununda "yoxdur" olaraq göstərildiyini görürük. Datanın daha təmiz olması üçün Çıxarış sütununu saxlayıb bill_of_sale sütununu silirik.

In [14]:
df.drop(columns=['bill_of_sale'], inplace=True)

Repair sütunu isə təmir sütununun ingiliscə adlandırılan versiyasıdır, hər iki sütun əmlakın təmirli olub-olmamasını göstərir. Sütunlara daha dəqiq baxsaq:

In [15]:
df['repair'].value_counts(dropna=False)

,count
repair,
Təmirli,81612
NaN,19163


In [16]:
df['Təmir'].value_counts(dropna=False)

,count
Təmir,
var,81612
yoxdur,13162
NaN,6001


Dəyərlərə baxdıqda repair sütununda "NaN" olaraq qeyd olunanların, Təmir sütununda "yoxdur" olaraq göstərildiyini görürük. Datanın daha təmiz olması üçün Təmir sütununu saxlayıb repair sütununu silirik.

In [17]:
df.drop(columns=['repair'], inplace=True)

Eyni şəkildə "mortgage" sütunu da "İpoteka" sütununun ingiliscə adlandırılmış halıdır və hər ikisi ipotekanın olub-olmamasını göstərir:

In [18]:
df['mortgage'].value_counts(dropna=False)

,count
mortgage,
NaN,67836
İpoteka var,32939


In [19]:
df['İpoteka'].value_counts(dropna=False)

,count
İpoteka,
NaN,67836
var,32939


Hər iki sütunda dəyərlər demək olar ki, eynidir. "mortgage" sütununu silib "İpoteka" sütununu saxlayırıq.

In [20]:
df.drop(columns=['mortgage'], inplace=True)

Duplicate sütunlar təmizləndikdən sonra sütun adlarını düzgün hala salaq.

In [21]:
df.rename(columns={
    'id_x': 'Elan_ID_1',
    'rel_url': 'Elan_URL',
    'estate_rel_url_x': 'Əmlak_URL',
    'datetime_scrape_x': 'Toplanma_tarixi_1',
    'price': 'Qiymət',
    'currency_x': 'Valyuta_1',
    'location': 'Məkan',
    'attributes': 'Əmlak_xüsusiyyətləri',
    'city': 'Şəhər',
    'day_x': 'Gün_1',
    'hour_x': 'Saat_1',
    'vip': 'VIP',
    'featured': 'Önə_çəkilmiş',
    'products_label': 'Elanı_yerləşdirən',
    'mortgage': 'İpoteka2',
    'img_url': 'Şəkil_URL',
    'id_y': 'Elan_ID_2',
    'estate_id': 'Əmlak_ID',
    'datetime_scrape_y': 'Toplanma_tarixi_2',
    'description': 'Təsvir',
    'unit_price': 'm²_qiyməti',
    'total_price': 'Ümumi_qiymət',
    'currency_y': 'Valyuta_2',
    'owner_name': 'Sahibin_adı',
    'owner_title': 'Sahibin_statusu',
    'shop_name': 'Agentliyin_adı',
    'shop_title': 'Agentliyin_növü',
    'address': 'Ünvan',
    'lat': 'Coğrafi_enlik',
    'lng': 'Coğrafi_uzunluq',
    'updated': 'Yenilənmə',
    'views': 'Baxış_sayı',
    'day_y': 'Gün_2',
    'hour_y': 'Saat_2',
    'estate_details_id_x': 'Detal_ID',
    'extra_info': 'Əlavə_məlumat',
    'Sahə': 'Sahə(m²)',
    'Torpaq sahəsi': 'Torpaq sahəsi(sot)'
}, inplace=True)

In [22]:
df.columns

Index(['Elan_ID_1', 'Elan_URL', 'Əmlak_URL', 'Toplanma_tarixi_1', 'Qiymət',
       'Valyuta_1', 'Məkan', 'Əmlak_xüsusiyyətləri', 'Şəhər', 'Gün_1',
       'Saat_1', 'VIP', 'Önə_çəkilmiş', 'Elanı_yerləşdirən', 'Şəkil_URL',
       'Elan_ID_2', 'Əmlak_ID', 'Toplanma_tarixi_2', 'Təsvir', 'm²_qiyməti',
       'Ümumi_qiymət', 'Valyuta_2', 'Sahibin_adı', 'Sahibin_statusu',
       'Agentliyin_adı', 'Agentliyin_növü', 'Ünvan', 'Coğrafi_enlik',
       'Coğrafi_uzunluq', 'Yenilənmə', 'Baxış_sayı', 'Gün_2', 'Saat_2',
       'Detal_ID', 'Binanın növü', 'Kateqoriya', 'Mərtəbə', 'Otaq sayı',
       'Sahə(m²)', 'Torpaq sahəsi(sot)', 'Təmir', 'Çıxarış', 'İpoteka',
       'Əlavə_məlumat'],
      dtype='object')

Sütunlar daha aydın şəkildə əks olundu.

# Data tiplərinin yoxlanılması

In [23]:
df.dtypes
#data tiplərinin yoxlanması

,0
Elan_ID_1,object
Elan_URL,object
Əmlak_URL,object
Toplanma_tarixi_1,object
Qiymət,float64
Valyuta_1,object
Məkan,object
Əmlak_xüsusiyyətləri,object
Şəhər,object
Gün_1,object


İlk öncə tarix sütunlarını "datetime" formatına çeviririk:

In [24]:
df['Toplanma_tarixi_1'] = (
    pd.to_datetime(df['Toplanma_tarixi_1'], format='mixed', utc=True)
      .dt.tz_localize(None)
)

In [25]:
df['Toplanma_tarixi_2'] = (
    pd.to_datetime(df['Toplanma_tarixi_2'], format='mixed', utc=True)
      .dt.tz_localize(None)
)

In [26]:
for col in ['Gün_1', 'Gün_2']:
    df[col] = pd.to_datetime(df[col], format='%d.%m.%Y')

In [27]:
for col in ['Saat_1', 'Saat_2']:
    df[col] = pd.to_datetime(df[col], format='%H:%M')

Otaq sayı sütunu "float" olaraq verilib, lakin "int" tipində olmalıdır. Otaq sayı natural ədəd olduğu üçün onu int tipinə çevirək:

In [28]:
df['Otaq sayı'] = df['Otaq sayı'].astype('Int64')

Sahə sütunu "object" olaraq verilib, bunun səbəbi sahələrin "m²" ilə birlikdə yazılmasıdır, sütunu "float" tipinə çevirək:

In [29]:
df['Sahə(m²)'] = np.where(
    df['Sahə(m²)'].str.contains('sot', case=False, na=False),
    df['Sahə(m²)'].str.extract(r'(\d+(?:\.\d+)?)')[0].astype(float) * 100,
    df['Sahə(m²)'].str.extract(r'(\d+(?:\.\d+)?)')[0].astype(float)
)

Sahə sütununu "float" tipinə çevirmək üçün m² hissələrini kəsərkən sahə sütununda "sot" ifadəsinin də yer aldığını müşahidə etdik və sot=100 m² olduğunu nəzərə alaraq sot ilə ifadə olunan sahələri m²-a çevirib sonra float tipinə çevirdik.

Torpaq sahəsi sütunu "object" olaraq verilib, bunun səbəbi torpaq sahəsinin "sot" ifadəsi ilə birlikdə verilməsidir, sütunu "float" tipinə çevirək:

In [30]:
df['Torpaq sahəsi(sot)'] = (
    df['Torpaq sahəsi(sot)']
    .str.replace(' sot', '', regex=False)
    .astype('float')
)

Torpaq sahəsi sütunundakı "sot" ifadəsini kəsib float tipinə çevirdik.

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100775 entries, 0 to 100774
Data columns (total 44 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   Elan_ID_1             100775 non-null  object        
 1   Elan_URL              100775 non-null  object        
 2   Əmlak_URL             100775 non-null  object        
 3   Toplanma_tarixi_1     100775 non-null  datetime64[ns]
 4   Qiymət                100775 non-null  float64       
 5   Valyuta_1             100775 non-null  object        
 6   Məkan                 100775 non-null  object        
 7   Əmlak_xüsusiyyətləri  100775 non-null  object        
 8   Şəhər                 100775 non-null  object        
 9   Gün_1                 100775 non-null  datetime64[ns]
 10  Saat_1                100775 non-null  datetime64[ns]
 11  VIP                   8596 non-null    object        
 12  Önə_çəkilmiş          3149 non-null    object        
 13 

Bütün data tipləri düzgün hala gətirildi.

# Dublikatların yoxlanılması

In [32]:
df.duplicated().sum()
#dublikat dəyərlərin yoxlanması

np.int64(0)

Heçbir dublikat sətir yoxdur.

# Boş dəyərlərin yoxlanması və idarə olunması

In [33]:
df.isnull().mean()*100
#null (boş) dəyərlərin yoxlanması

,0
Elan_ID_1,0.000000
Elan_URL,0.000000
Əmlak_URL,0.000000
Toplanma_tarixi_1,0.000000
Qiymət,0.000000
Valyuta_1,0.000000
Məkan,0.000000
Əmlak_xüsusiyyətləri,0.000000
Şəhər,0.000000
Gün_1,0.000000


Əvvəlcə faizi çox yüksək (>80) sütunlara baxaq:

## `VIP` sütunu

In [34]:
df['VIP'].value_counts(dropna=False)

,count
VIP,
NaN,92179
vipped,8596


bill_of_sales sütununu sildikdə müşahidə etdik ki, çıxarışı olan əmlaklar qeyd olunub, olmayanlar isə NaN olub. Bunun səbəbi data toplanarkən çıxarışı olanların qeyd olunması, çıxarışı olmayanların isə boş saxlanılmasıdır. VIP və Çıxarış dəyişənləri ikili (binary) kateqorik dəyişənlərdir. Ona görə də vip elanlar qeyd olunub olmayanlar NaN olaraq düşüb.

In [35]:
df['VIP'] = df['VIP'].eq('vipped').astype(int)

In [36]:
df['VIP'].value_counts(dropna=False)

,count
VIP,
0,92179
1,8596


NaN dəyərləri 0 (vip olmayan və ya normal elan) olaraq doldurduq, boş dəyərlər 0-la əvəz olundu.

## `Önə_çəkilmiş` sütunu

In [37]:
df['Önə_çəkilmiş'].value_counts(dropna=False)

,count
Önə_çəkilmiş,
NaN,97626
featured,3149


VIP sütununda olduğu kimi Önə_çəkilmiş sütunundakı dəyişənlər də binary kateqorik dəyişənlərdir. Çünki 2 seçim var: ya elan önə çəkilir (elan siyahısında daha yuxarı və ya xüsusi bölmədə göstərilir) ya da normal elan olur.

In [38]:
df['Önə_çəkilmiş'] = df['Önə_çəkilmiş'].eq('featured').astype(int)

In [39]:
df['Önə_çəkilmiş'].value_counts(dropna=False)

,count
Önə_çəkilmiş,
0,97626
1,3149


NaN dəyərləri 0 (önə çəkilməmiş və ya normal elan) olaraq doldurduq, boş dəyərlər 0-la əvəz olundu.

## `Binanın növü` sütunu

In [40]:
df['Binanın növü'].value_counts(dropna=False)

,count
Binanın növü,
NaN,100621
Ev / Mənzil,85
Biznes mərkəzi,64
Villa,5


NaN dəyərləri mümkün qədər azaltmağa çalışırıq, bunun üçün əvvəlcə kateqoriya sütunundan istifadə edirik. Ev/Mənzil bina növünə əsasən mənzil çoxmərtəbəli binada yerləşən yaşayış sahəsi olduğu üçün Həyət evi/Bağ evi kateqoriyasına aid bina növündəki dəyərləri (NaN) villa olaraq əvəz edirik:

In [41]:
mask = (
    df['Binanın növü'].isna() &
    (df['Kateqoriya'] == 'Həyət evi/Bağ evi')
)

df.loc[mask, 'Binanın növü'] = 'Villa'

In [42]:
df['Binanın növü'].value_counts(dropna=False)
# boş dəyərlərin yenidən yoxlanılması

,count
Binanın növü,
NaN,85050
Villa,15576
Ev / Mənzil,85
Biznes mərkəzi,64


Kateqoriya sütunundan yenidən istifadə edərək "obyekt" kateqoriyasına aid bina növlərini (NaN olanları) "Biznes mərkəzi" olaraq əvəz edirik:

In [43]:
mask = (
    df['Binanın növü'].isna() &
    (df['Kateqoriya'] == 'Obyekt')
)

df.loc[mask, 'Binanın növü'] = 'Biznes mərkəzi'

In [44]:
df['Binanın növü'].value_counts(dropna=False)
# boş dəyərlərin yenidən yoxlanılması

,count
Binanın növü,
NaN,81004
Villa,15576
Biznes mərkəzi,4110
Ev / Mənzil,85


Növbəti mərhələdə otaq sayına əsasən boş dəyərləri doldururuq:

In [45]:
df.loc[df['Binanın növü'] == 'Ev / Mənzil', 'Otaq sayı'].max()

np.int64(15)

In [46]:
df.loc[df['Binanın növü'] == 'Ev / Mənzil', 'Otaq sayı'].mean()

np.float64(4.588235294117647)

In [47]:
df.loc[df['Binanın növü'] == 'Biznes mərkəzi', 'Otaq sayı'].max()

np.int64(20)

In [48]:
df.loc[df['Binanın növü'] == 'Biznes mərkəzi', 'Otaq sayı'].mean()

np.float64(6.328125)

Ev / Mənzil və Biznes mərkəzi üçün dəyərləri tamamlayırıq. Hər iki bina növü üçün max və mean dəyərlər araşdırılır və alınan nəticələrə əsasən boş dəyərlər müvafiq olaraq doldurulur:

In [49]:
mask = df['Binanın növü'].isna()

df.loc[
    mask & (df['Otaq sayı'] >= 15),
    'Binanın növü'
] = 'Biznes mərkəzi'

df.loc[
    mask & (df['Otaq sayı'] < 15),
    'Binanın növü'
] = 'Ev / Mənzil'

In [50]:
df['Binanın növü'].value_counts(dropna=False)
# boş dəyərlərin yenidən yoxlanılması

,count
Binanın növü,
Ev / Mənzil,76076
Villa,15576
NaN,5008
Biznes mərkəzi,4115


Qalan boş dəyərləri "Digər" olaraq doldururuq:

In [51]:
df['Binanın növü'] = df['Binanın növü'].fillna('Digər')

In [52]:
df['Binanın növü'].value_counts(dropna=False)
# boş dəyərlərin yenidən yoxlanılması

,count
Binanın növü,
Ev / Mənzil,76076
Villa,15576
Digər,5008
Biznes mərkəzi,4115


Bütün boş dəyərlər müvafiq olaraq dolduruldu.

## `Mərtəbə` sütunu

In [53]:
df['Mərtəbə'].isna().value_counts()

,count
Mərtəbə,
False,75996
True,24779


False → Mərtəbə məlumatı var.

True → Mərtəbə məlumatı yoxdur (NaN).

In [54]:
df['Mərtəbə'] = df['Mərtəbə'].fillna('Məlum deyil')

Mərətəbə sütunundakı boş dəyərləri doldurmaq üçün hər hansı qanunauyğunluq mövcud deyil, ona görə də "məlum deyil" olaraq əvəz edirik.

## `İpoteka` sütunu

In [55]:
df['İpoteka'].value_counts(dropna=False)

,count
İpoteka,
NaN,67836
var,32939


İpoteka sütunundakı dəyişənlər də binary kateqorik dəyişənlərdir. Ona görə də NaN olan dəyərləri "yoxdur" ilə əvəz edirik:

In [56]:
df['İpoteka'] = df['İpoteka'].fillna('yoxdur')

## `Elanı_yerləşdirən` sütunu

In [57]:
df['Elanı_yerləşdirən'].value_counts(dropna=False)

,count
Elanı_yerləşdirən,
Agentlik,71981
NaN,28145
Kompleks,649


Elanı_yerləşdirən sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Məlum deyil" olaraq əvəz edirik.

In [58]:
df['Elanı_yerləşdirən'] = df['Elanı_yerləşdirən'].fillna('Məlum deyil')

## `Saat_2` sütunu

In [59]:
df['Saat_2'].isna().value_counts(dropna=False)

,count
Saat_2,
True,85144
False,15631


False → Saat məlumatı var.

True → Saat məlumatı yoxdur (NaN).

Saat_2 sütunu əmlakın update olunmuş məlumatlarının daxil edildiyi saatlardır. Bu məlumatları silmək bizə heç bir data itkisi yaratmır, çünki analiz zamanı məlumatların update olunmuş halını istifadə edirik, nə vaxt update olunduğu önəmli data deyil.

In [60]:
df.drop(columns=['Saat_2'], inplace=True)

## `Təsvir` sütunu

In [61]:
df['Təsvir'].isna().value_counts(dropna=False)

,count
Təsvir,
False,100510
True,265


False → Təsvir məlumatı var.

True → Təsvir məlumatı yoxdur (NaN).

Təsvir sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Əlavə olunmayıb" olaraq əvəz edirik.

In [62]:
df['Təsvir'] = df['Təsvir'].fillna('Əlavə olunmayıb')

## `m²_qiyməti` sütunu

In [63]:
df['m²_qiyməti'].isna().value_counts(dropna=False)

,count
m²_qiyməti,
False,75996
True,24779


False → m² qiyməti məlumatı var.

True → m² qiyməti məlumatı yoxdur (NaN).

Boş dəyərləri sütunun öz qiymətlərini istifadə edərək doldururuq:

In [64]:
qiymet_df = df[['m²_qiyməti']].copy()

qiymet_df['m²_qiyməti'] = (
    qiymet_df['m²_qiyməti']
    .str.replace(' AZN/m²', '', regex=False)
    .str.replace(' ', '', regex=False)
    .astype(float)
)

median = qiymet_df['m²_qiyməti'].median()
print(f"Ortalama qiymət: {median}")

Ortalama qiymət: 2360.0


Sütun daxilində text olduğu üçün obyekt tipindədir, ayrıca bir sütun halına gətirib data tipini dəyişdikdən sonra median ilə ortalama qiymət tapılır. Qiymətlər arasında outlier qiymətlər olduğu üçün ortalama qiyməti tapmaq üçün mean deyil mediandan istifadə edirik.

In [65]:
df['m²_qiyməti'] = df['m²_qiyməti'].fillna(median)

## `Torpaq sahəsi(sot)` sütunu

In [66]:
df['Torpaq sahəsi(sot)'].isna().value_counts(dropna=False)

,count
Torpaq sahəsi(sot),
True,85204
False,15571


False → Torpaq sahəsi məlumatı var.

True → Torpaq sahəsi məlumatı yoxdur (NaN).

Torpaq sahəsi(sot) sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Məlum deyil" olaraq əvəz edirik.

In [67]:
df['Torpaq sahəsi(sot)'] = (df['Torpaq sahəsi(sot)'].astype('object').fillna('Məlum deyil'))

Sütun float data tipində idi, "məlum deyil" ilə əvəz etmək üçün əvvəlcə data tipini object etdik, sonra isə boş dəyərləri doldurduq.

## `Sahibin_adı` sütunu

In [68]:
df['Sahibin_adı'].isna().value_counts(dropna=False)

,count
Sahibin_adı,
False,100118
True,657


False → Sahibin adı məlumatı var.

True → Sahibin adı məlumatı yoxdur (NaN).

Sahibin adı sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Məlum deyil" olaraq əvəz edirik.

In [69]:
df['Sahibin_adı'] = (df['Sahibin_adı'].fillna('Məlum deyil'))

## `Sahibin_statusu` sütunu

In [70]:
df['Sahibin_statusu'].value_counts(dropna=False)

,count
Sahibin_statusu,
vasitəçi (agent),89733
mülkiyyətçi,10385
NaN,657


Sahibin statusu sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Digər" olaraq əvəz edirik.

In [71]:
df['Sahibin_statusu'] = (df['Sahibin_statusu'].fillna('Digər'))

## `Agentliyin_adı` sütunu

In [72]:
df['Agentliyin_adı'].isna().value_counts(dropna=False)

,count
Agentliyin_adı,
False,72018
True,28757


False → Agentliyin adı məlumatı var.

True → Agentliyin adı məlumatı yoxdur (NaN).

Agentliyin adı sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Məlum deyil" olaraq əvəz edirik.




In [73]:
df['Agentliyin_adı'] = (df['Agentliyin_adı'].fillna('Məlum deyil'))

## `Agentliyin_növü` sütunu

In [74]:
df['Agentliyin_növü'].value_counts(dropna=False)

,count
Agentliyin_növü,
Daşınmaz əmlak agentliyi,72018
NaN,28757


Agentliyin növü sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Digər" olaraq əvəz edirik.

In [75]:
df['Agentliyin_növü'] = (df['Agentliyin_növü'].fillna('Digər'))

## `Təmir` sütunu

In [76]:
df['Təmir'].value_counts(dropna=False)

,count
Təmir,
var,81612
yoxdur,13162
NaN,6001


Təmir sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Məlum deyil" olaraq əvəz edirik.

In [77]:
df['Təmir'] = (df['Təmir'].fillna('Məlum deyil'))

## `Otaq sayı` sütunu

In [78]:
df['Otaq sayı'].isna().value_counts(dropna=False)

,count
Otaq sayı,
False,91363
True,9412


False → Təmir məlumatı var.

True → Təmir məlumatı yoxdur (NaN).

Otaq sayı sütunundakı boş dəyərləri doldurmaq üçün hər hansı əlavə məlumat yoxdur, ona görə də "Məlum deyil" olaraq əvəz edirik.

In [79]:
df['Otaq sayı'] = (df['Otaq sayı'].astype('object').fillna('Məlum deyil'))

In [80]:
df.isnull().sum()

,0
Elan_ID_1,0
Elan_URL,0
Əmlak_URL,0
Toplanma_tarixi_1,0
Qiymət,0
Valyuta_1,0
Məkan,0
Əmlak_xüsusiyyətləri,0
Şəhər,0
Gün_1,0
